File: ios/Sources/ILDSDesignSystem/IldsTextArea.swift
import ILDSTokens
import SwiftUI

// Mirrors lib/ilds_text_area.dart + web TextArea.tsx. TODO(web/desktop): hover states — mobile-first defer.

public struct IldsTextArea: View {
    @Binding private var text: String
    private let label: String?
    private let placeholder: String?
    private let helperText: String?
    private let errorText: String?
    private let successText: String?
    private let minRows: Int
    private let maxLength: Int?
    private let showCharCount: Bool
    private let isDisabled: Bool
    private let isReadOnly: Bool
    private let isLoading: Bool
    private let onChange: ((String) -> Void)?

    @FocusState private var isFocused: Bool

    public init(
        text: Binding<String>,
        label: String? = nil,
        placeholder: String? = nil,
        helperText: String? = nil,
        errorText: String? = nil,
        successText: String? = nil,
        minRows: Int = 3,
        maxLength: Int? = nil,
        showCharCount: Bool = false,
        isDisabled: Bool = false,
        isReadOnly: Bool = false,
        isLoading: Bool = false,
        onChange: ((String) -> Void)? = nil
    ) {
        _text = text
        self.label = label
        self.placeholder = placeholder
        self.helperText = helperText
        self.errorText = errorText
        self.successText = successText
        self.minRows = minRows
        self.maxLength = maxLength
        self.showCharCount = showCharCount
        self.isDisabled = isDisabled
        self.isReadOnly = isReadOnly
        self.isLoading = isLoading
        self.onChange = onChange
    }

    private var hasError: Bool { errorText != nil }
    private var hasSuccess: Bool { successText != nil }
    private var isInteractive: Bool { !isDisabled && !isLoading }

    public var body: some View {
        let colors = IldsTextAreaColors.resolve(
            hasError: hasError,
            hasSuccess: hasSuccess,
            isFocused: isFocused,
            isDisabled: isDisabled,
            isReadOnly: isReadOnly
        )
        let bottomText = errorText ?? successText ?? helperText
        let showCounter = showCharCount && maxLength != nil

        VStack(alignment: .leading, spacing: ILDSTokens.sp4) {
            if let label {
                Text(label)
                    .font(.system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightMedium))
                    .foregroundStyle(ILDSTokens.neutralCoolgray500)
            }

            ZStack(alignment: .topTrailing) {
                TextEditor(text: $text)
                    .font(.system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightRegular))
                    .foregroundStyle(colors.text)
                    .scrollContentBackground(.hidden)
                    .padding(.horizontal, ILDSTokens.sp16)
                    .padding(.vertical, ILDSTokens.sp12)
                    .frame(minHeight: CGFloat(minRows) * 24)
                    .focused($isFocused)
                    .disabled(!isInteractive || isReadOnly)
                    .onChange(of: text) { newValue in
                        if let maxLength, newValue.count > maxLength {
                            text = String(newValue.prefix(maxLength))
                        }
                        onChange?(text)
                    }
                    .overlay(alignment: .topLeading) {
                        if text.isEmpty, let placeholder {
                            Text(placeholder)
                                .font(.system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightRegular))
                                .foregroundStyle(ILDSTokens.neutralCoolgray300)
                                .padding(.horizontal, ILDSTokens.sp16 + 4)
                                .padding(.vertical, ILDSTokens.sp12 + 8)
                                .allowsHitTesting(false)
                        }
                    }

                if isLoading {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(ILDSTokens.primaryOrange500)
                        .frame(width: ILDSTokens.sp20, height: ILDSTokens.sp20)
                        .padding(.top, ILDSTokens.sp12)
                        .padding(.trailing, ILDSTokens.sp12)
                }
            }
            .background(colors.background)
            .overlay {
                RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium)
                    .strokeBorder(colors.border, lineWidth: colors.borderWidth)
            }
            .clipShape(RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium))

            if bottomText != nil || showCounter {
                HStack(alignment: .top) {
                    if let bottomText {
                        Text(bottomText)
                            .font(.system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightRegular))
                            .foregroundStyle(colors.bottomText)
                    }
                    Spacer()
                    if showCounter, let maxLength {
                        Text("\(text.count)/\(maxLength)")
                            .font(.system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightRegular))
                            .foregroundStyle(ILDSTokens.neutralCoolgray400)
                    }
                }
            }
        }
        .accessibilityLabel(label ?? placeholder ?? "Text area")
    }
}

private struct IldsTextAreaColors {
    let border: Color
    let background: Color
    let text: Color
    let bottomText: Color
    let borderWidth: CGFloat

    static func resolve(
        hasError: Bool,
        hasSuccess: Bool,
        isFocused: Bool,
        isDisabled: Bool,
        isReadOnly: Bool
    ) -> IldsTextAreaColors {
        let border: Color
        if isDisabled {
            border = ILDSTokens.neutralCoolgray200
        } else if hasError {
            border = ILDSTokens.errorRed600
        } else if isFocused {
            border = ILDSTokens.primaryOrange500
        } else if hasSuccess {
            border = ILDSTokens.successGreen600
        } else {
            border = ILDSTokens.neutralCoolgray300
        }

        let background: Color = isDisabled || isReadOnly
            ? ILDSTokens.neutralCoolgray50
            : ILDSTokens.globalWhite000

        let text: Color
        if isDisabled {
            text = ILDSTokens.neutralCoolgray300
        } else if isReadOnly {
            text = ILDSTokens.neutralCoolgray500
        } else {
            text = ILDSTokens.neutralCoolgray900
        }

        let bottomText: Color
        if hasError {
            bottomText = ILDSTokens.errorRed600
        } else if hasSuccess {
            bottomText = ILDSTokens.successGreen600
        } else {
            bottomText = ILDSTokens.neutralCoolgray400
        }

        let borderWidth: CGFloat = hasError || isFocused ? 2 : 1
        return IldsTextAreaColors(
            border: border,
            background: background,
            text: text,
            bottomText: bottomText,
            borderWidth: borderWidth
        )
    }
}

File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsTextArea.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Mirrors web/TextArea.tsx + lib/ilds_text_area.dart.

@Composable
fun IldsTextArea(
    modifier: Modifier = Modifier,
    value: String = "",
    onValueChange: (String) -> Unit = {},
    label: String? = null,
    placeholder: String? = null,
    helperText: String? = null,
    errorText: String? = null,
    successText: String? = null,
    minLines: Int = 3,
    maxLength: Int? = null,
    showCharCount: Boolean = false,
    isDisabled: Boolean = false,
    isReadOnly: Boolean = false,
    isLoading: Boolean = false,
) {
    var text by remember(value) { mutableStateOf(value) }
    val interactionSource = remember { androidx.compose.foundation.interaction.MutableInteractionSource() }
    val isFocused by interactionSource.collectIsFocusedAsState()
    val hasError = errorText != null
    val hasSuccess = successText != null && !hasError
    val interactive = !isDisabled && !isLoading

    val borderColor = when {
        isDisabled -> IldsTokens.neutralCoolgray200
        hasError -> IldsTokens.errorRed600
        isFocused -> IldsTokens.primaryOrange500
        hasSuccess -> IldsTokens.successGreen500
        else -> IldsTokens.neutralCoolgray300
    }
    val borderWidth = if (hasError || isFocused) 2.dp else 1.dp
    val fillColor = when {
        isDisabled || isReadOnly -> IldsTokens.neutralCoolgray50
        else -> IldsTokens.globalWhite000
    }
    val textColor = when {
        isDisabled -> IldsTokens.neutralCoolgray300
        isReadOnly -> IldsTokens.neutralCoolgray500
        else -> IldsTokens.neutralCoolgray900
    }
    val bottomText = errorText ?: successText ?: helperText
    val bottomColor = when {
        hasError -> IldsTokens.errorRed600
        hasSuccess -> IldsTokens.successGreen600
        else -> IldsTokens.neutralCoolgray400
    }

    Column(
        modifier = modifier.semantics {
            contentDescription = label ?: placeholder ?: "Text area"
        },
        verticalArrangement = Arrangement.spacedBy(IldsTokens.sp4),
    ) {
        if (label != null) {
            Text(
                text = label,
                style = TextStyle(
                    fontSize = IldsTokens.fontSize12,
                    fontWeight = IldsTokens.fontWeightMedium,
                ),
                color = IldsTokens.neutralCoolgray500,
            )
        }
        Surface(
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(IldsTokens.radiusMedium),
            color = fillColor,
            border = BorderStroke(borderWidth, borderColor),
        ) {
            Box(modifier = Modifier.defaultMinSize(minHeight = 128.dp)) {
                BasicTextField(
                    value = text,
                    onValueChange = {
                        val next = if (maxLength != null) it.take(maxLength) else it
                        text = next
                        onValueChange(next)
                    },
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(
                            horizontal = IldsTokens.sp16,
                            vertical = IldsTokens.sp12,
                        ),
                    enabled = interactive,
                    readOnly = isReadOnly,
                    minLines = minLines,
                    textStyle = TextStyle(
                        fontSize = IldsTokens.fontSize14,
                        lineHeight = 18.sp,
                        fontWeight = IldsTokens.fontWeightRegular,
                        color = textColor,
                    ),
                    cursorBrush = SolidColor(IldsTokens.primaryOrange500),
                    interactionSource = interactionSource,
                    decorationBox = { inner ->
                        Box {
                            if (text.isEmpty() && placeholder != null) {
                                Text(
                                    text = placeholder,
                                    style = TextStyle(
                                        fontSize = IldsTokens.fontSize14,
                                        lineHeight = 18.sp,
                                        fontWeight = IldsTokens.fontWeightRegular,
                                    ),
                                    color = IldsTokens.neutralCoolgray300,
                                )
                            }
                            inner()
                        }
                    },
                )
                if (isLoading) {
                    CircularProgressIndicator(
                        modifier = Modifier
                            .align(androidx.compose.ui.Alignment.TopEnd)
                            .padding(IldsTokens.sp12)
                            .size(20.dp),
                        color = IldsTokens.primaryOrange500,
                        strokeWidth = 2.dp,
                    )
                }
            }
        }
        if (bottomText != null || (showCharCount && maxLength != null)) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                if (bottomText != null) {
                    Text(
                        text = bottomText,
                        modifier = Modifier.weight(1f),
                        style = TextStyle(
                            fontSize = IldsTokens.fontSize12,
                            lineHeight = 16.sp,
                            fontWeight = IldsTokens.fontWeightRegular,
                        ),
                        color = bottomColor,
                    )
                } else {
                    Box(modifier = Modifier.weight(1f))
                }
                if (showCharCount && maxLength != null) {
                    Text(
                        text = "${text.length}/$maxLength",
                        style = TextStyle(
                            fontSize = IldsTokens.fontSize12,
                            fontWeight = IldsTokens.fontWeightRegular,
                        ),
                        color = IldsTokens.neutralCoolgray400,
                    )
                }
            }
        }
    }
}
